File: reactive_autonomous_nav/astar_planner.py
#!/usr/bin/env python3
"""
A* Global Planner — reactive_autonomous_nav
  • Octile heuristic (tight for 8-connected grids, no sqrt)
  • Collision-checked Laplacian path smoothing
  • Rich RViz markers: explored heat-map, start/goal spheres, glowing path
"""
print(''.join(chr(x-7) for x in [104,105,107,124,115,39,121,104,111,116,104,117]))

import rclpy
import rclpy.time
import heapq
import math
import numpy as np

from rclpy.node             import Node
from rclpy.duration         import Duration
from rclpy.qos              import QoSProfile, DurabilityPolicy, ReliabilityPolicy
from tf2_ros                import TransformListener, Buffer
from nav_msgs.msg           import OccupancyGrid, Path
from geometry_msgs.msg      import PoseStamped, Point
from std_msgs.msg           import String, ColorRGBA
from visualization_msgs.msg import Marker, MarkerArray

# ── cost thresholds ──────────────────────────────────────────────────
LETHAL_COST = 253
AVOID_COST  = 200
FREE_COST   = 10

# ── replan gates ─────────────────────────────────────────────────────
PATH_BLOCKED_LOOKAHEAD = 2.0   # m — check this far ahead for obstacles
MAX_PATH_DEVIATION     = 0.8   # m — replan if robot this far from path
MIN_REPLAN_DISTANCE    = 0.3   # m — don't replan if robot hasn't moved


class AStarPlannerNode(Node):

    def __init__(self):
        super().__init__('astar_planner_node')

        # ── state ────────────────────────────────────────────────────
        self.global_data   = None
        self.global_info   = None
        self.global_origin = None

        self.local_data    = None
        self.local_info    = None
        self.local_origin  = None

        self.odom_to_map       = None
        self.last_goal         = None
        self.current_path      = None
        self.goal_reached      = False
        self.last_replan_pose  = None

        # ── TF ───────────────────────────────────────────────────────
        self.tf_buffer   = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)

        cmap_qos = QoSProfile(
            depth       = 1,
            durability  = DurabilityPolicy.TRANSIENT_LOCAL,
            reliability = ReliabilityPolicy.RELIABLE,
        )

        # ── subs ─────────────────────────────────────────────────────
        self.create_subscription(OccupancyGrid, '/global_costmap/costmap',
                                 self._global_cb,  cmap_qos)
        self.create_subscription(OccupancyGrid, '/local_costmap/costmap',
                                 self._local_cb,   cmap_qos)
        self.create_subscription(PoseStamped,   '/goal_pose',
                                 self._goal_cb,    10)
        self.create_subscription(String,        '/replan_request',
                                 self._replan_cb,  10)
        self.create_subscription(String,        '/dwa_status',
                                 self._status_cb,  10)

        # ── pubs ─────────────────────────────────────────────────────
        self.path_pub      = self.create_publisher(Path,        '/plan',             10)
        self.status_pub    = self.create_publisher(String,      '/astar_status',     10)
        self.explored_pub  = self.create_publisher(MarkerArray, '/astar_explored',   10)
        self.marker_pub    = self.create_publisher(MarkerArray, '/astar_markers',    10)

        # ── timer ────────────────────────────────────────────────────
        self.create_timer(0.5, self._check_and_replan)

        self.get_logger().info('A* Global Planner — ready')

    # ================================================================
    #  Callbacks
    # ================================================================
    def _global_cb(self, msg: OccupancyGrid):
        self.global_data   = np.array(msg.data, dtype=np.int16).reshape(
            (msg.info.height, msg.info.width))
        self.global_info   = msg.info
        self.global_origin = (msg.info.origin.position.x,
                              msg.info.origin.position.y)
        self.get_logger().info(
            f'Global costmap: {msg.info.width}×{msg.info.height} '
            f'res={msg.info.resolution:.3f}',
            throttle_duration_sec=10.0)

    def _local_cb(self, msg: OccupancyGrid):
        self.local_data   = np.array(msg.data, dtype=np.int16).reshape(
            (msg.info.height, msg.info.width))
        self.local_info   = msg.info
        self.local_origin = (msg.info.origin.position.x,
                             msg.info.origin.position.y)

    def _status_cb(self, msg: String):
        if msg.data == 'REACHED':
            self.goal_reached = True
            self.last_goal    = None
            self.get_logger().info('Goal reached — stopping replanning')
            
    def _replan_cb(self, msg: String):
        """Force replan from DWA — skip all gates."""
        if self.last_goal is not None:
            self.get_logger().info('Force replan from DWA')
            self._plan(self.last_goal)

    def _goal_cb(self, msg: PoseStamped):
        self.last_goal        = msg
        self.goal_reached     = False
        self.current_path     = None
        self.last_replan_pose = None
        self._plan(msg)

    # ================================================================
    #  TF helpers
    # ================================================================
    def _get_odom_to_map(self):
        try:
            t = self.tf_buffer.lookup_transform(
                'map', 'odom', rclpy.time.Time(), Duration(seconds=0.2))
            tr  = t.transform.translation
            rot = t.transform.rotation
            yaw = math.atan2(2.0 * (rot.w * rot.z + rot.x * rot.y),
                             1.0 - 2.0 * (rot.y**2 + rot.z**2))
            self.odom_to_map = (tr.x, tr.y, yaw)
        except Exception:
            pass

    def _get_robot_pose(self):
        try:
            t = self.tf_buffer.lookup_transform(
                'map', 'base_link', rclpy.time.Time(), Duration(seconds=0.5))
            return t.transform.translation.x, t.transform.translation.y
        except Exception as e:
            self.get_logger().warn(f'TF lookup failed: {e}')
            return None

    # ================================================================
    #  Costmap helpers
    # ================================================================
    def _local_cost_at_map(self, wx_map, wy_map) -> int:
        if self.local_data is None or self.odom_to_map is None:
            return -1
        tmx, tmy, tyaw = self.odom_to_map
        dx = wx_map - tmx
        dy = wy_map - tmy
        wx_odom =  math.cos(tyaw) * dx + math.sin(tyaw) * dy
        wy_odom = -math.sin(tyaw) * dx + math.cos(tyaw) * dy
        res = self.local_info.resolution
        col = int((wx_odom - self.local_origin[0]) / res)
        row = int((wy_odom - self.local_origin[1]) / res)
        if not (0 <= col < self.local_info.width and
                0 <= row < self.local_info.height):
            return -1
        return int(self.local_data[row, col])

    def _w2g(self, wx, wy):
        res = self.global_info.resolution
        col = int((wx - self.global_origin[0]) / res)
        row = int((wy - self.global_origin[1]) / res)
        return row, col

    def _g2w(self, row, col):
        res = self.global_info.resolution
        return ((col + 0.5) * res + self.global_origin[0],
                (row + 0.5) * res + self.global_origin[1])

    def _in_bounds(self, row, col):
        return (0 <= row < self.global_info.height and
                0 <= col < self.global_info.width)

    def _global_cell_cost(self, row, col) -> int:
        if not self._in_bounds(row, col):
            return 255
        v = int(self.global_data[row, col])
        return v if v >= 0 else 255

    def _merged_cell_cost(self, row, col) -> int:
        global_c = self._global_cell_cost(row, col)
        wx, wy   = self._g2w(row, col)
        local_c  = self._local_cost_at_map(wx, wy)
        if local_c < 0:
            return global_c
        return max(global_c, local_c)

    def _is_traversable(self, row, col) -> bool:
        return self._merged_cell_cost(row, col) < LETHAL_COST

    def _traversal_cost(self, cell_c: int) -> float:
        if cell_c <= FREE_COST:
            return 1.0
        if cell_c < AVOID_COST:
            return 1.0 + (cell_c - FREE_COST) / (AVOID_COST - FREE_COST) * 49.0
        return 500.0

    # ================================================================
    #  Replan gating
    # ================================================================
    def _is_path_blocked(self, rx, ry) -> bool:
        if self.current_path is None or self.local_data is None:
            return False
        for ps in self.current_path.poses:
            wp = ps.pose.position
            d  = math.hypot(wp.x - rx, wp.y - ry)
            if d > PATH_BLOCKED_LOOKAHEAD:
                continue
            row, col = self._w2g(wp.x, wp.y)
            c = self._merged_cell_cost(row, col)
            if c >= LETHAL_COST:
                self.get_logger().info(
                    f'Path blocked at ({wp.x:.2f},{wp.y:.2f}) cost={c}')
                return True
        return False

    def _path_deviation(self, rx, ry) -> float:
        if self.current_path is None:
            return 0.0
        min_d = float('inf')
        for ps in self.current_path.poses:
            wp = ps.pose.position
            d  = math.hypot(wp.x - rx, wp.y - ry)
            if d < min_d:
                min_d = d
        return min_d

    def _robot_has_moved(self, rx, ry) -> bool:
        if self.last_replan_pose is None:
            return True
        lx, ly = self.last_replan_pose
        return math.hypot(rx - lx, ry - ly) >= MIN_REPLAN_DISTANCE

    def _check_and_replan(self):
        if self.last_goal is None or self.goal_reached:
            return
        if self.global_data is None or self.current_path is None:
            return
        self._get_odom_to_map()
        pose = self._get_robot_pose()
        if pose is None:
            return
        rx, ry = pose

        if self._is_path_blocked(rx, ry):
            self.get_logger().info('Replan: path blocked by local obstacle')
            self._plan(self.last_goal)
            return

        dev = self._path_deviation(rx, ry)
        if dev > MAX_PATH_DEVIATION:
            if self._robot_has_moved(rx, ry):
                self.get_logger().info(
                    f'Replan: deviation {dev:.2f}m > {MAX_PATH_DEVIATION}m')
                self._plan(self.last_goal)
            return

        self.get_logger().debug(
            'Path valid — no replan needed', throttle_duration_sec=2.0)

    # ================================================================
    #  A* core — octile heuristic
    # ================================================================
    def _astar(self, start, goal):
        MOVES = [
            (-1,  0, 1.0), ( 1, 0, 1.0), (0, -1, 1.0), (0, 1, 1.0),
            (-1, -1, 1.414), (-1, 1, 1.414), (1, -1, 1.414), (1, 1, 1.414),
        ]
        open_set = []
        heapq.heappush(open_set, (0.0, start))
        came_from = {}
        g = {start: 0.0}
        explored = []

        while open_set:
            _, current = heapq.heappop(open_set)
            explored.append(current)

            if current == goal:
                path = []
                while current in came_from:
                    path.append(current)
                    current = came_from[current]
                path.append(start)
                return list(reversed(path)), explored

            for dr, dc, move_cost in MOVES:
                nb = (current[0] + dr, current[1] + dc)
                if not self._is_traversable(*nb):
                    continue
                cell_c = self._merged_cell_cost(*nb)
                ng     = g[current] + move_cost * self._traversal_cost(cell_c)
                if nb not in g or ng < g[nb]:
                    came_from[nb] = current
                    g[nb] = ng
                    # ── octile heuristic (no sqrt) ───────────────────
                    dx = abs(nb[0] - goal[0])
                    dy = abs(nb[1] - goal[1])
                    h  = max(dx, dy) + 0.4142135 * min(dx, dy)
                    heapq.heappush(open_set, (ng + h, nb))

        return None, explored

    # ================================================================
    #  Collision-checked Laplacian smoothing
    # ================================================================
    def _smooth(self, path_world, iterations=50):
        if len(path_world) < 3:
            return path_world
        smooth = [list(p) for p in path_world]
        wd, ws = 0.5, 0.3
        for _ in range(iterations):
            for i in range(1, len(smooth) - 1):
                candidate = [smooth[i][0], smooth[i][1]]
                for j in range(2):
                    candidate[j] += wd * (path_world[i][j] - candidate[j])
                    candidate[j] += ws * (smooth[i-1][j] + smooth[i+1][j]
                                          - 2.0 * candidate[j])
                # ── collision guard: only accept if still traversable ──
                row, col = self._w2g(candidate[0], candidate[1])
                if self._merged_cell_cost(row, col) < LETHAL_COST:
                    smooth[i] = candidate
        return [tuple(p) for p in smooth]

    # ================================================================
    #  RViz marker helpers
    # ================================================================
    def _publish_explored(self, explored, path_cells):
        """Explored-cell cubes with a blue→cyan heat-map gradient."""
        path_set = set(map(tuple, path_cells)) if path_cells else set()
        rejected = [c for c in explored if tuple(c) not in path_set]
        step     = max(1, len(rejected) // 1200)
        rejected = rejected[::step]

        now = self.get_clock().now().to_msg()
        ma  = MarkerArray()

        # single CUBE_LIST marker is far cheaper than N individual cubes
        m = Marker()
        m.header.frame_id = 'map'
        m.header.stamp    = now
        m.ns     = 'explored'
        m.id     = 0
        m.type   = Marker.CUBE_LIST
        m.action = Marker.ADD
        res = self.global_info.resolution
        m.scale.x = res
        m.scale.y = res
        m.scale.z = 0.01  # thin wafer

        n = max(len(rejected), 1)
        for idx, (row, col) in enumerate(rejected):
            wx, wy = self._g2w(row, col)
            m.points.append(Point(x=wx, y=wy, z=0.02))
            # gradient from deep blue (early) → cyan (late)
            t = idx / n
            m.colors.append(ColorRGBA(
                r=0.0,
                g=0.2 + 0.6 * t,
                b=0.9 - 0.3 * t,
                a=0.25 + 0.15 * t,
            ))

        ma.markers.append(m)
        self.explored_pub.publish(ma)

    def _publish_start_goal_markers(self, rx, ry, gx, gy):
        """Bright spheres on start & goal."""
        now = self.get_clock().now().to_msg()
        ma  = MarkerArray()

        for i, (x, y, r, g, b, label) in enumerate([
            (rx, ry, 0.2, 0.8, 1.0, 'start'),
            (gx, gy, 1.0, 0.3, 0.3, 'goal'),
        ]):
            m = Marker()
            m.header.frame_id = 'map'
            m.header.stamp    = now
            m.ns = label
            m.id = i
            m.type   = Marker.SPHERE
            m.action = Marker.ADD
            m.pose.position.x    = x
            m.pose.position.y    = y
            m.pose.position.z    = 0.25
            m.pose.orientation.w = 1.0
            m.scale.x = 0.3
            m.scale.y = 0.3
            m.scale.z = 0.3
            m.color = ColorRGBA(r=r, g=g, b=b, a=0.9)
            ma.markers.append(m)

        self.marker_pub.publish(ma)

    # ================================================================
    #  Plan
    # ================================================================
    def _plan(self, goal_msg: PoseStamped):
        if self.global_data is None:
            self.get_logger().warn('Cannot plan: global_data is None')
            return
        self._get_odom_to_map()
        if self.odom_to_map is None:
            self.get_logger().warn('Cannot plan: odom_to_map is None')
        pose = self._get_robot_pose()
        if pose is None:
            self.get_logger().warn('Cannot plan: _get_robot_pose returned None')
            return
        rx, ry = pose
        gx = goal_msg.pose.position.x
        gy = goal_msg.pose.position.y

        if math.hypot(gx - rx, gy - ry) < 0.15:
            return

        start = self._w2g(rx, ry)
        goal  = self._w2g(gx, gy)

        if not self._is_traversable(*goal):
            s = 'UNREACHABLE: Goal inside obstacle'
            self.get_logger().warn(s)
            self.status_pub.publish(String(data=s))
            return

        if not self._is_traversable(*start):
            found = False
            for dr in range(-5, 6):
                for dc in range(-5, 6):
                    ns = (start[0] + dr, start[1] + dc)
                    if self._is_traversable(*ns):
                        start = ns
                        found = True
                        break
                if found:
                    break

        path_cells, explored = self._astar(start, goal)

        # ── RViz: explored heat-map + start/goal spheres ─────────────
        self._publish_explored(explored, path_cells)
        self._publish_start_goal_markers(rx, ry, gx, gy)

        if path_cells is None:
            s = 'UNREACHABLE: A* found no path'
            self.get_logger().warn(s)
            self.status_pub.publish(String(data=s))
            return

        path_world = [self._g2w(r, c) for r, c in path_cells]
        path_world = self._smooth(path_world)

        # ── publish nav_msgs/Path ────────────────────────────────────
        path_msg = Path()
        path_msg.header.frame_id = 'map'
        path_msg.header.stamp    = self.get_clock().now().to_msg()
        for wx, wy in path_world:
            ps = PoseStamped()
            ps.header = path_msg.header
            ps.pose.position.x    = wx
            ps.pose.position.y    = wy
            ps.pose.orientation.w = 1.0
            path_msg.poses.append(ps)
        self.path_pub.publish(path_msg)

        self.current_path     = path_msg
        self.last_replan_pose = (rx, ry)
        self.get_logger().info(
            f'Path: {len(path_msg.poses)} wps '
            f'({rx:.2f},{ry:.2f})→({gx:.2f},{gy:.2f})')


def main(args=None):
    rclpy.init(args=args)
    node = AStarPlannerNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == '__main__':
    main()

File: cpp/src/astar_planner.cpp
// astar_planner.cpp — A* Global Planner (C++)
// reactive_nav_cpp | ROS2 Jazzy
//
// Octile heuristic (tight for 8-connected grids)
// Collision-checked Laplacian path smoothing
// RViz markers: explored frontier + glowing path

#include <rclcpp/rclcpp.hpp>
#include <nav_msgs/msg/occupancy_grid.hpp>
#include <nav_msgs/msg/path.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <geometry_msgs/msg/point.hpp>
#include <visualization_msgs/msg/marker.hpp>
#include <visualization_msgs/msg/marker_array.hpp>
#include <tf2_ros/transform_listener.h>
#include <tf2_ros/buffer.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>

#include <algorithm>
#include <cmath>
#include <limits>
#include <queue>
#include <vector>

// ── cost thresholds ──────────────────────────────────────────────────────────
static constexpr uint8_t LETHAL_COST           = 253;
static constexpr float   PATH_BLOCKED_LOOKAHEAD = 2.0f;   // m
static constexpr float   MAX_PATH_DEVIATION     = 0.8f;   // m
static constexpr float   MIN_REPLAN_DISTANCE    = 0.3f;   // m

// 8-connected grid neighbors
static constexpr int   DX[8]        = { 1,-1, 0, 0, 1,-1, 1,-1};
static constexpr int   DY[8]        = { 0, 0, 1,-1, 1,-1,-1, 1};
static constexpr float STEP_COST[8] = {1.0f,1.0f,1.0f,1.0f,1.414f,1.414f,1.414f,1.414f};

struct AStarNode {
    float f, g;
    int   idx;
    bool operator>(const AStarNode& o) const { return f > o.f; }
};

// ── helpers ──────────────────────────────────────────────────────────────────
static inline float octile(int dx, int dy)
{
    int a = std::abs(dx), b = std::abs(dy);
    if (a < b) std::swap(a, b);
    return static_cast<float>(a) + 0.4142f * static_cast<float>(b);
}

static inline uint8_t cell_cost(const nav_msgs::msg::OccupancyGrid& map, int idx)
{
    return static_cast<uint8_t>(map.data[idx]);
}

// ── node ─────────────────────────────────────────────────────────────────────
class AStarPlannerNode : public rclcpp::Node
{
public:
    AStarPlannerNode() : Node("astar_planner_node")
    {
        auto qos_map = rclcpp::QoS(1).transient_local().reliable();

        sub_global_ = create_subscription<nav_msgs::msg::OccupancyGrid>(
            "/global_costmap/costmap", qos_map,
            [this](nav_msgs::msg::OccupancyGrid::SharedPtr m){ global_map_ = m; });

        sub_local_ = create_subscription<nav_msgs::msg::OccupancyGrid>(
            "/local_costmap/costmap", qos_map,
            [this](nav_msgs::msg::OccupancyGrid::SharedPtr m){ local_map_ = m; });

        sub_goal_ = create_subscription<geometry_msgs::msg::PoseStamped>(
            "/goal_pose", 10,
            std::bind(&AStarPlannerNode::on_goal, this, std::placeholders::_1));

        pub_plan_    = create_publisher<nav_msgs::msg::Path>("/global_plan", 10);
        pub_markers_ = create_publisher<visualization_msgs::msg::MarkerArray>(
                           "/astar/markers", 10);

        tf_buffer_   = std::make_shared<tf2_ros::Buffer>(get_clock());
        tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);

        replan_timer_ = create_wall_timer(
            std::chrono::milliseconds(500),
            std::bind(&AStarPlannerNode::check_replan, this));

        RCLCPP_INFO(get_logger(), "A* Planner (C++) ready");
    }

private:
    // ── state ─────────────────────────────────────────────────────────────────
    nav_msgs::msg::OccupancyGrid::SharedPtr  global_map_;
    nav_msgs::msg::OccupancyGrid::SharedPtr  local_map_;
    nav_msgs::msg::Path::SharedPtr           current_path_;
    geometry_msgs::msg::PoseStamped::SharedPtr last_goal_;
    double last_replan_x_{0}, last_replan_y_{0};
    bool   goal_reached_{false};
    bool   has_replan_pose_{false};

    // ── ROS ───────────────────────────────────────────────────────────────────
    rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::SharedPtr  sub_global_;
    rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::SharedPtr  sub_local_;
    rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr sub_goal_;
    rclcpp::Publisher<nav_msgs::msg::Path>::SharedPtr              pub_plan_;
    rclcpp::Publisher<visualization_msgs::msg::MarkerArray>::SharedPtr pub_markers_;
    rclcpp::TimerBase::SharedPtr                                   replan_timer_;
    std::shared_ptr<tf2_ros::Buffer>                               tf_buffer_;
    std::shared_ptr<tf2_ros::TransformListener>                    tf_listener_;

    // ── TF ────────────────────────────────────────────────────────────────────
    bool get_robot_pose(double& rx, double& ry) const
    {
        try {
            auto tf = tf_buffer_->lookupTransform(
                "map", "base_link", tf2::TimePointZero, tf2::durationFromSec(0.3));
            rx = tf.transform.translation.x;
            ry = tf.transform.translation.y;
            return true;
        } catch (...) { return false; }
    }

    // ── grid utils ────────────────────────────────────────────────────────────
    bool map_free(const nav_msgs::msg::OccupancyGrid& map, int x, int y) const
    {
        const auto& info = map.info;
        if (x < 0 || y < 0 || x >= (int)info.width || y >= (int)info.height) return false;
        return cell_cost(map, y * (int)info.width + x) < LETHAL_COST;
    }

    std::pair<int,int> world_to_grid(const nav_msgs::msg::OccupancyGrid& map,
                                     double wx, double wy) const
    {
        float res = map.info.resolution;
        return {
            static_cast<int>((wx - map.info.origin.position.x) / res),
            static_cast<int>((wy - map.info.origin.position.y) / res)
        };
    }

    std::pair<double,double> grid_to_world(const nav_msgs::msg::OccupancyGrid& map,
                                           int gx, int gy) const
    {
        float res = map.info.resolution;
        return {
            map.info.origin.position.x + (gx + 0.5) * res,
            map.info.origin.position.y + (gy + 0.5) * res
        };
    }

    // ── A* core ───────────────────────────────────────────────────────────────
    std::vector<std::pair<int,int>> run_astar(
        const nav_msgs::msg::OccupancyGrid& map,
        int sx, int sy, int gx, int gy) const
    {
        const int W = (int)map.info.width, H = (int)map.info.height;
        if (sx < 0 || sy < 0 || sx >= W || sy >= H) return {};
        if (gx < 0 || gy < 0 || gx >= W || gy >= H) return {};

        const int total    = W * H;
        const int goal_idx = gy * W + gx;
        const int start    = sy * W + sx;

        std::vector<float> g_cost(total, std::numeric_limits<float>::infinity());
        std::vector<int>   parent(total, -1);
        std::vector<bool>  closed(total, false);

        g_cost[start] = 0.0f;

        std::priority_queue<AStarNode, std::vector<AStarNode>, std::greater<AStarNode>> open;
        open.push({octile(gx - sx, gy - sy), 0.0f, start});

        while (!open.empty()) {
            auto [f, g, cur] = open.top(); open.pop();
            if (closed[cur]) continue;
            closed[cur] = true;
            if (cur == goal_idx) break;

            int cx = cur % W, cy = cur / W;
            for (int d = 0; d < 8; d++) {
                int nx = cx + DX[d], ny = cy + DY[d];
                if (!map_free(map, nx, ny)) continue;
                int nidx = ny * W + nx;
                if (closed[nidx]) continue;

                // weight by cell cost so planner naturally avoids high-cost cells
                float step_g = STEP_COST[d] * (1.0f + cell_cost(map, nidx) / 255.0f);
                float ng     = g + step_g;
                if (ng < g_cost[nidx]) {
                    g_cost[nidx] = ng;
                    parent[nidx] = cur;
                    open.push({ng + octile(gx - nx, gy - ny), ng, nidx});
                }
            }
        }

        // backtrack
        std::vector<std::pair<int,int>> path;
        int cur = goal_idx;
        while (cur != -1 && cur != start) {
            path.emplace_back(cur % W, cur / W);
            cur = parent[cur];
        }
        if (cur == start) path.emplace_back(sx, sy);
        std::reverse(path.begin(), path.end());
        return path;
    }

    // ── Laplacian smoothing ────────────────────────────────────────────────────
    std::vector<std::pair<int,int>> smooth_path(
        std::vector<std::pair<int,int>> path,
        const nav_msgs::msg::OccupancyGrid& map,
        int iters = 5, float alpha = 0.5f) const
    {
        for (int it = 0; it < iters; it++) {
            for (size_t i = 1; i + 1 < path.size(); i++) {
                int nx = static_cast<int>((path[i-1].first  + path[i+1].first)  / 2.0f * alpha
                                        + path[i].first  * (1.0f - alpha));
                int ny = static_cast<int>((path[i-1].second + path[i+1].second) / 2.0f * alpha
                                        + path[i].second * (1.0f - alpha));
                if (map_free(map, nx, ny)) path[i] = {nx, ny};
            }
        }
        return path;
    }

    // ── planning ─────────────────────────────────────────────────────────────
    void plan_to_goal(const geometry_msgs::msg::PoseStamped& goal)
    {
        if (!global_map_) { RCLCPP_WARN(get_logger(), "No costmap yet"); return; }
        double rx, ry;
        if (!get_robot_pose(rx, ry)) { RCLCPP_WARN(get_logger(), "TF not ready"); return; }

        auto [sx, sy] = world_to_grid(*global_map_, rx, ry);
        auto [gx, gy] = world_to_grid(*global_map_, goal.pose.position.x, goal.pose.position.y);

        auto raw      = run_astar(*global_map_, sx, sy, gx, gy);
        if (raw.empty()) { RCLCPP_WARN(get_logger(), "A* found no path"); return; }
        auto smoothed = smooth_path(raw, *global_map_);

        nav_msgs::msg::Path plan;
        plan.header.frame_id = "map";
        plan.header.stamp    = now();
        for (auto& [px, py] : smoothed) {
            auto [wx, wy] = grid_to_world(*global_map_, px, py);
            geometry_msgs::msg::PoseStamped ps;
            ps.header              = plan.header;
            ps.pose.position.x     = wx;
            ps.pose.position.y     = wy;
            ps.pose.orientation.w  = 1.0;
            plan.poses.push_back(ps);
        }

        pub_plan_->publish(plan);
        current_path_     = std::make_shared<nav_msgs::msg::Path>(plan);
        goal_reached_     = false;
        has_replan_pose_  = false;

        RCLCPP_INFO(get_logger(), "Plan published (%zu waypoints)", plan.poses.size());
        publish_path_markers(plan);
    }

    void on_goal(geometry_msgs::msg::PoseStamped::SharedPtr msg)
    {
        last_goal_    = msg;
        goal_reached_ = false;
        plan_to_goal(*msg);
    }

    void check_replan()
    {
        if (!last_goal_ || goal_reached_ || !current_path_ || !global_map_) return;
        double rx, ry;
        if (!get_robot_pose(rx, ry)) return;

        double to_goal = std::hypot(
            rx - last_goal_->pose.position.x,
            ry - last_goal_->pose.position.y);
        if (to_goal < 0.3) { goal_reached_ = true; RCLCPP_INFO(get_logger(), "Goal reached"); return; }

        // don't replan if robot hasn't moved far enough
        if (has_replan_pose_) {
            if (std::hypot(rx - last_replan_x_, ry - last_replan_y_) < MIN_REPLAN_DISTANCE) return;
        }

        // check if robot drifted off path
        float min_d = std::numeric_limits<float>::max();
        for (auto& ps : current_path_->poses) {
            float d = (float)std::hypot(rx - ps.pose.position.x, ry - ps.pose.position.y);
            if (d < min_d) min_d = d;
        }
        if (min_d > MAX_PATH_DEVIATION) {
            RCLCPP_INFO(get_logger(), "Replanning — deviation %.2f m", min_d);
            last_replan_x_   = rx;
            last_replan_y_   = ry;
            has_replan_pose_ = true;
            plan_to_goal(*last_goal_);
        }
    }

    // ── RViz markers ──────────────────────────────────────────────────────────
    void publish_path_markers(const nav_msgs::msg::Path& plan) const
    {
        visualization_msgs::msg::MarkerArray ma;

        visualization_msgs::msg::Marker line;
        line.header    = plan.header;
        line.ns        = "astar_path";
        line.id        = 0;
        line.type      = visualization_msgs::msg::Marker::LINE_STRIP;
        line.action    = visualization_msgs::msg::Marker::ADD;
        line.scale.x   = 0.05f;
        line.color.r   = 0.0f;
        line.color.g   = 1.0f;
        line.color.b   = 0.4f;
        line.color.a   = 1.0f;
        for (auto& ps : plan.poses) {
            geometry_msgs::msg::Point p;
            p.x = ps.pose.position.x;
            p.y = ps.pose.position.y;
            line.points.push_back(p);
        }
        ma.markers.push_back(line);

        // start sphere
        visualization_msgs::msg::Marker start_m;
        start_m.header  = plan.header;
        start_m.ns      = "astar_start";
        start_m.id      = 1;
        start_m.type    = visualization_msgs::msg::Marker::SPHERE;
        start_m.action  = visualization_msgs::msg::Marker::ADD;
        start_m.scale.x = start_m.scale.y = start_m.scale.z = 0.2;
        start_m.color.r = 0.2f; start_m.color.g = 0.8f; start_m.color.b = 0.2f; start_m.color.a = 1.0f;
        if (!plan.poses.empty()) start_m.pose = plan.poses.front().pose;
        ma.markers.push_back(start_m);

        // goal sphere
        visualization_msgs::msg::Marker goal_m = start_m;
        goal_m.ns     = "astar_goal";
        goal_m.id     = 2;
        goal_m.color.r = 1.0f; goal_m.color.g = 0.2f; goal_m.color.b = 0.2f;
        if (!plan.poses.empty()) goal_m.pose = plan.poses.back().pose;
        ma.markers.push_back(goal_m);

        pub_markers_->publish(ma);
    }
};

int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<AStarPlannerNode>());
    rclcpp::shutdown();
    return 0;
}
